lib/validation.js

```diff
@@ -76,32 +76,43 @@ function validateParam (validatorFunction, request, paramName) {
 
 function validate (context, request) {
   var params = validateParam(context[paramsSchema], request, 'params')
+  var schemaErrorFormatter = (context.schemaErrorFormatter || schemaErrorsText).bind(request)
+
   if (params) {
-    return wrapValidationError(params, 'params')
+    return wrapValidationError(params, 'params', schemaErrorFormatter)
   }
   var body = validateParam(context[bodySchema], request, 'body')
   if (body) {
-    return wrapValidationError(body, 'body')
+    return wrapValidationError(body, 'body', schemaErrorFormatter)
   }
   var query = validateParam(context[querystringSchema], request, 'query')
   if (query) {
-    return wrapValidationError(query, 'querystring')
+    return wrapValidationError(query, 'querystring', schemaErrorFormatter)
   }
   var headers = validateParam(context[headersSchema], request, 'headers')
   if (headers) {
-    return wrapValidationError(headers, 'headers')
+    return wrapValidationError(headers, 'headers', schemaErrorFormatter)
   }
   return null
 }
 
-function wrapValidationError (result, dataVar) {
+function wrapValidationError (result, dataVar, schemaErrorFormatter) {
   if (result instanceof Error) {
     return result
   }
-  var error = new Error(schemaErrorsText(result, dataVar))
-  error.validation = result
-  error.validationContext = dataVar
-  return error
+
+  const validationError = new Error()
+
+  try {
+    validationError.message = schemaErrorFormatter(result, dataVar)
+  } catch (err) {
+    validationError.formatterError = `schema custom formatter error: ${err.message}`
+    return validationError
+  }
+
+  validationError.validation = result
+  validationError.validationContext = dataVar
+  return validationError
```

Comment:
I would implement this differently. I think the `errorFormatter` should return the `error` itself. The default `errorFormatter` should be the one we did before.

Reply:
I am confused about the need for `validationError.formatterError` here 🤔

Reply:
There is none.

Reply:
We need to differentiate between the errors, if we just return the `err` object we won't be able to tell the difference and will return `400`, this is for marking this error as an exception inside the `schemaErrorFormatter` so it would be possible to return `500`.

Reply:
If there is a better way to do this, and make the request return 500, please tell!

@mcollina I don't exactly understand what you mean.